sway-core/src/ir_generation/function.rs

```diff
@@ -2727,6 +2731,110 @@ impl<'a> FnCompiler<'a> {
+                ..
+            } => self.collect_event_metadata_for_type(context, md_mgr, *type_id, span)?,
+            TypeInfo::Struct(decl_ref) => {
+                let decl = decl_engine.get_struct(decl_ref);
+                if decl.attributes.event().is_none() {
+                    None
+                } else {
+                    let indexed_fields: Vec<_> = decl
+                        .fields
+                        .iter()
+                        .take_while(|field| field.attributes.indexed().is_some())
+                        .collect();
+
+                    let indexed_count = indexed_fields.len();
+                    let field_size = if let Some(first_field) = indexed_fields.first() {
+                        let size =
+                            self.indexed_field_size_in_bytes(context, md_mgr, first_field)?;
+                        for field in indexed_fields.iter().skip(1) {
+                            let current_size =
+                                self.indexed_field_size_in_bytes(context, md_mgr, field)?;
+                            if current_size != size {
+                                return Err(CompileError::Internal(
+                                    "Indexed event fields must have matching sizes.",
+                                    field.span.clone(),
+                                ));
+                            }
+                        }
+                        Some(size)
+                    } else {
+                        None
+                    };
+
+                    let count = u16::try_from(indexed_count).map_err(|_| {
+                        CompileError::Internal(
+                            "Too many indexed fields on event for current metadata format.",
+                            span.clone(),
+                        )
+                    })?;
+
+                    Some(sway_ir::LogEventData::for_event(field_size, count))
+                }
+            }
+            TypeInfo::Enum(decl_ref) => {
+                if decl_engine.get_enum(decl_ref).attributes.event().is_some() {
+                    Some(sway_ir::LogEventData::for_event(None, 0))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        })
+    }
+
+    fn indexed_field_size_in_bytes(
+        &mut self,
+        context: &mut Context,
+        md_mgr: &mut MetadataManager,
+        field: &ty::TyStructField,
+    ) -> Result<u8, CompileError> {
+        let ir_ty = convert_resolved_type_id(
+            self.engines,
+            context,
+            md_mgr,
+            self.module,
+            Some(self),
+            field.type_argument.type_id,
+            &field.span,
+        )?;
+        let size = ir_ty.size(context).in_bytes();
+        u8::try_from(size).map_err(|_| {
+            CompileError::Internal(
+                "Indexed event field size exceeds supported range.",
+                field.span.clone(),
```

Comment:
nit: improved error message

```suggestion
                &format!(
                    "Indexed event field size ({} bytes) exceeds maximum supported size ({} bytes).",
                    size, u8::MAX
                ),
                field.span.clone(),
```